File: Labs/Lab 6/Main.py
def canFinish(numCourses, prerequisites):
    G = {i:[] for i in range(numCourses)}
    for course, pre in prerequisites:
        G[course].append(pre)
    visited = [0]*numCourses
    def dfs(course):
        visited[course] = 1
        for pre in G[course]:
            if visited[pre] == 1: return True
            if visited[pre] == 0: 
                if dfs(pre): return True
        visited[course] = 2
        return False
    for course in range(numCourses):
        if visited[course] == 0:
            if dfs(course): return False
    return True

File: Labs/Lab 6/Main.java
import java.util.*; 
public class Main {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        ArrayList<Integer>[] graph = new ArrayList[numCourses];
        for(int i=0;i<numCourses;i++)
            graph[i] = new ArrayList<Integer>();
    
        boolean[] visited = new boolean[numCourses];
        boolean[] stack = new boolean[numCourses];
        for(int i=0; i<prerequisites.length;i++)
            graph[prerequisites[i][1]].add(prerequisites[i][0]);
    
        for(int i=0; i<numCourses; i++) {
            if(!visited[i] && isCyclic(i, visited, stack, graph)) 
                return false;
        }
        return true;
    }
    private boolean isCyclic(int v, boolean[] visited, boolean[] stack, ArrayList<Integer>[] graph) {
        if(stack[v])
            return true;
        if(visited[v])
            return false;
        visited[v] = true;
        stack[v] = true;
        for(int i=0; i<graph[v].size(); i++) {
            if(isCyclic(graph[v].get(i), visited, stack, graph))
                return true;
        }
        stack[v] = false;
        return false;
    }
}